### DISCUSSION TITLE
Gambit Compiler's CLI for Modular Programming

### DISCUSSION THREAD

**Discussion by feeley** [gambit/gambit#887] [Ideas]
# Gambit Compiler's CLI for Modular Programming

Gambit supports modular programming using R7RS style modules as well as "primitive" modules. Similarly to the programming model used by C and other languages, each module can be compiled separately and then linked into an executable program.

Unfortunately, the Gambit Scheme Compiler `gsc` currently has a CLI that is rather low level and not ideal for building modular programs. To build a modular program the programmer must be aware of many details, such as the dependencies between the modules, their location in the filesystem, and which ones need to be rebuilt.

This document explains the design of a high level CLI for the compiler. We welcome constructive feedback on this proposal.

To make things concrete, let's assume we are interested in building two programs `A.exe` and `X.exe` with the following module dependencies, where `A` and `X` are the respective main modules of those programs:

```mermaid
graph
A((A)) --> B((B))
A --> L
L --> M
L((L)) --> N((N))
M((M)) --> O((O))
N((N)) --> O((O))
X((X)) --> L
X --> Y((Y))
```

Note that both programs need module `L` and its dependencies `M`, `N`, and `O`.

The source code for these modules can be expressed using R7RS `define-library` forms (typically in `.sld` files, `A.sld`, etc) or using explicit `##supply-module`/`##demand-module` forms (typically in `.scm` and `#.scm` files, `A.scm` and `A#.scm`, etc), or a mix. A layout for these files in the filesystem can use directories to group files that are related:

```
example
├── A.sld
├── B.sld
├── mylib
│   ├── L.sld
│   ├── M.sld
│   ├── N.sld
│   └── O
│       ├── O.sld
│       └── helper.c
└── prototype
    ├── X.sld
    └── Y.sld
```

Ideally it should be possible to build an executable program with `gsc` by specifying only the main module and letting the compiler determine what are the other modules that need to be compiled and linked. To build `A.exe` something like this should be sufficient:

    gsc -exe -o A.exe example/ example/mylib/ A

Note the use of the arguments `example/` and `example/mylib/` on the command line to tell `gsc` which directories it should search for the modules, for example finding `A`'s source code in `example/A.sld`, `O`'s source code in `example/mylib/O/O.sld`, etc. The system maintains a list of places to search for modules, the *module search order*, initially containing `~~lib/` and `~~userlib/`, which is extended by the directories on the command line (any argument ending in `/` or `.`). System modules are normally installed in `~~lib/` and user installed modules are in `~~userlib/`.

Given the dependency graph above it is straightforward for `gsc` to determine that to build program `A.exe` the modules `A`, `B`, `L`, `M`, `N`, and `O` must be compiled and linked, and to build program `X.exe` the modules `X`, `Y`, `L`, `M`, `N`, and `O` must be compiled and linked.

## Build Caches

In the above scenario it would be wasteful to compile modules `L`, `M`, `N`, and `O` more than once when `A.exe` and `X.exe` are built. This is a typical situation for general purpose libraries that are used by many programs. An important special case is the Gambit runtime library itself that should not be recompiled every time an executable program is built. Note that the dependency to the Gambit runtime library has not been made explicit in the module dependency graph above, but in principle all (or nearly all) modules depend on it.

The obvious solution to avoid redundant compilations is to introduce a caching mechanism that will preserve the artefacts resulting from a module's compilation for future builds that require those artefacts. A *build cache* is a directory that contains these artefacts. For every module that is compiled `gsc` must first generate a *target file*. For example, when using the C target, `gsc` compiles `L.sld` into the target file `L.c`. This file is then compiled using the C compiler (e.g. `cc`) to produce the file `L.o` which is one of the object files needed by the C linker to build `A.exe`. The file `L.c` contains various meta-information about module `L` (embedded in the file as C comments) and in particular the direct module dependencies of module `L`, i.e. the list of modules `(M N)`. Preserving this target file in the build cache is useful because it allows quickly determining a module's dependencies without having to (yet again) parse, expand and process the module's source code file `L.sld`. Moreover, the target file can be compiled by the C compiler into the *dynamically loadable object file* `L.o1` that the `gsi` interpreter uses when importing modules at run time. So the files `L.c`, `L.o`, and `L.o1` are important to preserve in the build cache as artefacts derived from `L.sld`:

```mermaid
graph LR
L.sld -.-> gsc1([gsc])
gsc1 -.-> L.c
subgraph build cache
L.c -.-> cc1([cc])
cc1 -.-> L.o
L.c -.-> cc2([cc])
cc2 -.-> L.o1
end
style gsc1 fill:#aaa
style cc1 fill:#aaa
style cc2 fill:#aaa
```

## Compilation Settings

How should the build cache be managed and where should it be stored?

A module's build cache can be located with the same search algorithm that is used to locate a module's source code by placing it right next to that module's source code. To allow different versions of Gambit to coexist on a system, the build caches for these different versions must not interfere. This can be solved by adding the Gambit version information to the name of the build cache. For example the module `L` whose source code is in `example/mylib/L.sld` will have the build cache directory `example/mylib/L~409005/` for Gambit v4.9.5 and `example/mylib/L~409006/` for Gambit v4.9.6.

There are other factors that influence the target file that is generated for a module and the associated artefacts. Clearly the `gsc` compilation target (`C`, `js`, etc) is an important factor. Gambit compilation options also affect the target file. A module can be compiled to include debugging information so that more precise error messages and backtraces can be generated, and a module can be compiled to unsafe code that avoids run time checks for increased execution speed. In the case of some targets like `C`, the operating system, processor architecture, and optimization level that is used, will impact the artefacts generated.

A specific combination of these factors is called a *compilation setting*. The build cache distinguishes these different settings by using subdirectories whose names specify the setting. For example the subdirectory `C~debug~unsafe/` could be used for the setting that combines the C target, the generation of debugging information, and the generation of unsafe code (note that the encoding of this information in the directory name is not yet finalized and may be inspired by s-expression syntax such as `{{target~^C}~{debug~@@t}~{safe~@@f}}/`) . Here is a possible state of the filesystem for module `L` built with Gambit v4.9.5 for the `C` and `js` targets with 3 different settings:

```
example
└── mylib
    ├── L.sld
    └── L~409005
        ├── C
        │   ├── L.c
        │   ├── L.o
        │   └── L.o1
        ├── C~debug
        │   ├── L.c
        │   ├── L.o
        │   └── L.o1
        └── js
            ├── L.js
            ├── L.o
            └── L.o1
```

## Command Line Interface

The current `gsc` allows various compilation options on the command line to select the compilation setting, for example `-debug` to generate debugging information and `-prelude '(declare (not safe))'` to generate unsafe code. Using compilation options in a high level CLI that allows modules to be specified on the command line brings about the issue of scope. For example what is the meaning of the following invocation of the compiler (disregarding the module search order issue):

    gsc -exe -debug A

Should it mean that debugging information is generated for module `A` only, or also all the direct and indirect dependencies (`B`, `L`, `M`, `N`, and `O`)? It is probably intuitive to use as a default behaviour that the compilation options apply to the module and its dependencies, but this is not always what is needed so there should be a way to allow the programmer to have a fine control over the scope of the compilation options.

We propose a new syntax to control the scope of compilation options. Balanced curly braces can appear on the command line to delimit the scope of the compilation options. Curly braces are used because they don't typically need to be escaped at the command line of shells. There are two types of scopes: *deep* scopes have the syntax `{ ... }` and *singular* (or *shallow*) scopes have the syntax `{1 ... }`. Any scope can be nested in a deep scope, but only singular scopes can be nested in a singular scope. The compilation options in a singular scope apply to the set of modules in that scope and nested scopes, and not their dependencies. For a deep scope the compilation options also apply to the dependencies, except those covered previously on the command line. For example:

```
gsc -exe {1 -debug M } L { -prelude '(declare (not safe))' A }
```

will compile module M with debugging information, modules L, N, and O with the default compilation options, and modules A and B with unsafe code generation. There is an implicit deep scope around all the command line options and compilation options apply from left to right within their scope, so the following are equivalent to the previous invocation of `gsc`:

```
gsc -exe { {1 -debug M } L { -prelude '(declare (not safe))' A } }
gsc -exe {1 -debug M } L -prelude '(declare (not safe))' A
```

This syntax also finds use for attaching linking information to individual modules or group of modules. For example, when a module is to be dynamically linked instead of statically linked, the `-dynamic-link` option can be used:

```
gsc -exe { -dynamic-link L } A   # A,B are statically linked, and L,M,N,O are dynamically linked
```

Moreover, when developing a program it is useful to force a module or group of modules to be rebuilt. A `-rebuild` option can be used to indicate that the module's build cache must be emptied before creating its artefacts from the source code of the module:

```
gsc -exe { -rebuild L } A  # L,M,N,O are forcibly compiled after emptying their build cache
```

**Comment by gambiteer** [comment]
> So the files `L.c`, `L.o`, and `L.o1` are important to preserve in the build cache as artefacts derived from `L.sld`:

I'm sorry, why do you need `L.o` as well as `L.o1`? So you can build statically linked binaries?

**Reply by feeley** [reply]
Yes the `L.o` file is used for static linking and `L.o1` is the shared library for execution time loading of the module. While they are derived from the same `L.c` file they are subtly different in ways that depend on the C compiler (typically the `L.o1` file needs to be compiled with `-shared`, `-fPIC` and possibly other system specific linker options).

**Comment by gambiteer** [comment]
> For example the module `L` whose source code is in `example/mylib/L.sld` will have the build cache directory `example/mylib/L~409005/` for Gambit v4.9.5 and `example/mylib/L~409006/` for Gambit v4.9.6.

What about someone who runs intermediate Github versions of Gambit?

**Reply by feeley** [reply]
Intermediate "versions" will use the same version identifier as the latest release. Doing otherwise would mean that for every new commit everything needs to be compiled. Typically this is not needed, so I prefer that the programmer explicitly ask for this with the `-rebuild` option.

**Comment by gambiteer** [comment]
I don't have a reasoned opinion about what to do with this issue.  What do systems like Racket do?

**Reply by feeley** [reply]
Someone that knows more about Racket should answer this.

**Reply by gambiteer** [reply]
If you aren't curious about whether other systems solve similar problems (or whether they've even found solutions to similar problems, or even have similar problems), that's OK.  I just tend to look around to see what other people have done before coming up my own solution if needed.

**Reply by feeley** [reply]
It is not that I am not curious, but rather that I haven't created executables with Racket so I'm not the best one to answer your question.

For creating executables with Racket it seems the best documentation is: https://docs.racket-lang.org/raco/exe.html . I don't think it tackles the same issues, i.e. supporting several target languages and supporting compilation options.

**Comment by lassik** [comment]
The `{ ... }` command line usage is at the limit of what is comfortable and idiomatic to express in shell syntax.

Would it be better to use an S-expression domain specific language for complex jobs? Something like:

```Scheme
(exe (with debug M)
     L
     (all-with (prelude (declare (not safe)))
               A))

(exe (all-with dynamic-link L)
     A)

(exe (all-with rebuild L)
     A)
```

**Reply by feeley** [reply]
Where would these S-expressions be put? I assume in a separate file because on the command line all the parentheses need to be escaped in most shells. The point of the proposed syntax is to allow quickly trying out variations of how an application is linked. Basically viewing the gsc CLI as a form of REPL. If the linking information is in a separate file that file needs to be edited before the linking is done. It also means that there is a discontinuity between linking with no options:

```
$ gsc -exe myapp
```

and linking with a small variation that requests debugging module M, which in the proposed syntax is:

```
$ gsc -exe { -debug M } myapp
```

but with S-expressions would be:

```
$ gsc -build-with foo.link
$ cat foo.link
(exe (all-with debug M)
     myapp)
```

And it also introduces a possible inconsistency between the CLI option syntax/semantics (like `-debug`, etc) and the S-expression options (`debug`, etc).

Note that complex scenarios that need to be repeated could be put in a shell script:

```
$ cat build-with-debugging.sh
#!/bin/sh
gsc -exe { -debug M } L -prelude "(declare (not safe))" myapp
```

Note also that the same syntax is expected to be usable for `gsi` to execute a Scheme program with some contextual variations, for example:

```
$ gsi { -debug M } myapp
```

Having to edit a file to vary how the interpreter is started would be cumbersome and a departure from standard unix practice (command line options to alter a program's behaviour).

**Reply by amirouche** [reply]
> The `{ ... }` command line usage is at the limit of what is comfortable and idiomatic to express in shell syntax.

I agree with @lassik. I disagree with the s-exp DSL for the time being at least. 

The focus should be to provide a CLI, say `gsu`, with it, it will be easy to do most common task from a sh repl. `gsc`, and `gsi` will still be available for advanced usage (such as caching intermediate objects).

What I have in mind looks like:

```
gsu repl [DIRECTORY ...]
gsu exec [--unsafe][--dev] [DIRECTORY ...] MYLIBRARY.SLD
gsu compile [--unsafe][--dev][--javascript] [DIRECTORY ...] MYLIBRARY.SLD
gsu check [--unsafe][--dev][--javascript] [DIRECTORY ...] MYLIBRARY.SLD
```

I have a prototype working with chibi scheme. I does not support static binaries.

Instead of `MYPROGRAM.SCM`, mind the fact the primary argument is `MYLIBRARY.SLD`:

- Because I am not sure what is the point of `myprogram.scm` given libraries;
- It helps to jump start into TDD with the help of `gsu check` if the code is already in a library;

Let me know what you think.

**Comment by wqy24** [comment]
IMO nesting in command line options is terrible. So I agree with the sexp configs.

Quickly trying out can be done with  `-config` and `-unconfig`
```
gsc -conf gambconf.foo -unconfig '(L prelude)' -config '(A (B static-link) (L (O debug)))' A.scm
```
It also allow us to write configs incrementally:
```
gscfg -f gambconf.foo -unconfig '(A (L (O debug)))' -config '(N (O debug))' -o gambconf.foo.debug_N-O
```
Explain:
* DSL here is different from @lassik's. A short explain: `(N ...)` means options for `N`. `(N (O ...))` means options for `O` as a dependency of `N`.
* `debug` means `-debug` in command line options. I suggest that most of the config symbols are just command-line options without `-`
* For options with details I suggest use vectors: `#(prelude (declare (not safe)))`.
* The whole usage of `gscfg` is shown above.

I don't know much about `gsc` options. Just take a look.
I just hope that we can continue working on this.
